# Lab 02 Solutions

## Part 1

### Assignment
Lets use our new control flow skills to compute probabilities and conditional probabilities.

Run the below code to create one vector that describes whether patients were assigned to treatment or control and a second vector describing whether the patient had a succussful or unsuccesful status.

In [2]:
draw_random_number = function(){
    return(runif(1))
}

##### Practice loop and conditional structure
After running the above code block, we can create a random number by running the following code ``draw_random_number()``
1. Run the code ``draw_random_number()``
2. Create and if/else statement that prints "Less than 1/2" when ``draw_random_number()`` is smaller than the value 0.5 and else prints "greater than or equal to 1/2" when ``draw_random_number()`` is 0.5 or larger. 
3. Create a variable ``category`` and assign this variable the value -1. 
4. Create an if/else if/else statement that assigns the variable category the value 1 when ``draw_random_number()`` is smaller than 0.25, the value 2 when ``draw_random_number()`` is greater than or equal to 0.25 and smaller than 0.50, the value 3 when ``draw_random_number()`` is greater than or equal to 0.50 and smaller than 0.75 and the value 4 otherwise. 
5. Create a vector random_values that is empty (i.e. ``random_values = c()``). 
6. Assign the variable ``value`` equal to ``draw_random_number()``. 
7. Create a while loop with the condition that ``value`` returns a value less than 0.5
   - Inside the while loop, in the code block, assign the variable ``value`` to a new random number ``draw_random_number()``
   - Inside the while loop, in the code block, append ``value`` to your vector ``random_values``
8. Create a for loop that iterates the code block you create for the while loop 10 times. 
9. What can you say about the vector ``random_values`` that would be produced from 7. versus 8.?

**Question 1**  

1. Run the code ``draw_random_number()``  

To run a function, I type the name of the function into the code cell followed by open and close parentheses `()`. These parentheses are important because that's how you signal to R that we want to *run* the function, not just find out more information about the function itself.

In [3]:
draw_random_number()

[1] 0.8403342

**Question 2**  

2. Create and if/else statement that prints "Less than 1/2" when ``draw_random_number()`` is smaller than the value 0.5 and else prints "greater than or equal to 1/2" when ``draw_random_number()`` is 0.5 or larger.   

I remember that the basic format of an if / else is the following:  
```
if ( logical_expression ) {
## what to do if expression inside parentheses is true

} else {
## what to do if expression inside parentheses is false
}
```
Where we put some expression in the place of `logical_expression` that will reduce to a `TRUE` or `FALSE`. For example, if `x = 3` and `logical_expression = x < 5`, then `logical_expression` will reduce to `TRUE` and R executes any lines of code contained within the first set of brackets `{}` relating to the `if` and not to the set of brackets relating to the `else`.

In [4]:
if (draw_random_number() < 0.5) {
    print('Less than 1/2')
} else { ## i know that, if the random number isn't less than 0.5, it must be greater than or equal to 0.5
    print('Greater than or equal to 1/2')
}

[1] "Greater than or equal to 1/2"


**Question 3**  

3. Create a variable ``category`` and assign this variable the value -1. 

In [5]:
category = -1

**Question 4**  

4. Create an if/else if/else statement that assigns the variable category the value 1 when ``draw_random_number()`` is smaller than 0.25, the value 2 when ``draw_random_number()`` is greater than or equal to 0.25 and smaller than 0.50, the value 3 when ``draw_random_number()`` is greater than or equal to 0.50 and smaller than 0.75 and the value 4 otherwise.  

I can see that I will need to make multiple comparisons about a random number that I draw. Each time I call the function `draw_random_number()`, it's like picking a new random number out of a hat. Imagine you were going to pick a random number out of a hat and have a friend ask a bunch of questions to try to guess what that number is. It doesn't make much sense and wouldn't be fair to your friend if, after each question they ask, you keep drawing out new numbers. You need to take that first number that you picked out of the hat and *save it*, so that each question your friend asks will relate to the *same* number. Much in the same way, we want to only call `draw_random_number()` to pick a random number *once*, save it, and then ask some questions about it using if/else if/else logic. We can save the output of a function directly to a variable to store it for later use. Let's see how that works.

In [6]:
## each time i call draw_random number(), i get a different number
for (i in 1:3) {
    print(draw_random_number())
}

[1] 0.8161634
[1] 0.8370712
[1] 0.9460999


In [7]:
## but if i store the output of draw_random_number() to a variable, that value stays the same as long 
## as i don't overwrite that variable in my code later on
random_number = draw_random_number()
for (i in 1:3) {
    print(random_number)
}

[1] 0.1770283
[1] 0.1770283
[1] 0.1770283


In [8]:
## we use our stored number to ask a series of questions using a more complex if/else structure
if (random_number < 0.25) {
    category = 1
} else if (random_number >= 0.25 && random_number < 0.5) {
    category = 2
} else if (random_number >= 0.5 && random_number < 0.75) {
    category = 3
} else {
    category = 4
}

Because I can see that the random number I saved is equal to $0.5256$, I would expect `category` to equal $3$:

In [9]:
print(category)

[1] 1


**Question 5**

5. Create a vector random_values that is empty (i.e. ``random_values = c()``). 

In [10]:
random_values = c()

**Question 6**  
6. Assign the variable ``value`` equal to ``draw_random_number()``.  

In [11]:
value = draw_random_number()

**Question 7**  

7. Create a while loop with the condition that ``value`` returns a value less than 0.5
   - Inside the while loop, in the code block, assign the variable ``value`` to a new random number ``draw_random_number()``
   - Inside the while loop, in the code block, append ``value`` to your vector ``random_values``  
   
I remember that the structure for a `while` loop is the following:  
```
while( logical_expression ) {
    ## do these lines of code only if logical_expression is true
    ## repeat these lines of code only when logical_expression becomes false
}
```  

Where, again, `logical_expression` is some expression that reduces to `TRUE` or `FALSE`. Imagine back to our example of drawing random numbers out of a hat. This question is asking us to essentially keep drawing random numbers out of this hat until we get one that's less than 0.5, meanwhile we're going to store each random number that we draw into some container so we can access them later.

In [12]:
while (value > 0.5) {
    value = draw_random_number()
    random_values = c(random_values, value) ## remember, when i append, im using a vector to update that vector
}
print(random_values)

[1] 0.8570794 0.3314840


**Question 8**  
8. Create a for loop that iterates the code block you create for the while loop 10 times.  

The structure of a `for` loop:  
```
for (new_counter_variable in some_vector) {
    ## code to execute on each iteration
}
```
Where, on each iteration, `new_counter_variable` takes on the value of each element in `some_vector`. The loop continues until there are no more elements left to iterate over in `some_vector`.  

So I'll take the code inside the while loop and put it inside a `for` loop that will iterate ten times. For the slot `some_vector`, I will use a vector of length ten with elements one through ten with the colon notation: `1:10`.

In [13]:
for (i in 1:10) {
    value = draw_random_number()
    random_values = c(random_values, value)
}
print(random_values)

 [1] 0.85707943 0.33148399 0.61637705 0.98683915 0.55735863 0.14475398
 [7] 0.75473417 0.73836509 0.04500539 0.65791411 0.19953055 0.52303394


**Question 9**  
9. What can you say about the vector ``random_values`` that would be produced from 7. versus 8.?  

With the `while` loop, I can see that how many times the loop will run will totally depend on the 'luck of the draw', meaning that the loop will only continue if the randomly-drawn number is greater than 0.5. The `for` loop will run for sure ten times no matter what.

## Part 2

##### Lets create a for loop that will iterate through the positions of our two vectors ``patient_status`` and ``patient_treatment``, performing different operations as we move to each item in the vector  
1. Use the ``length`` function to compute the number of patients in the study and store that length in the variable ``N``
2. Create a variable called ``num_of_treat_assigns`` and assign that variable the value 0
3. Create a variable called ``num_of_contr_assigns`` and assign that variable the value 0
4. Create a variable called ``num_of_status_success`` and assign that variable the value 0
5. Create a for loop that will run a code block starting at 1 and ending at ``N``
6. Inside the code block of the for loop, use the variable ``num_of_status_success`` to count the number of patients with a succussful status
7. Use ``N`` and ``num_of_status_success`` to compute the probability of a success
8. Lets make our for loop more complicated. Inside our code block, create an if/else such that when patient treatment is a "t" we increment the variable ``num_of_treat_assigns`` by 1, else we increment the variable ``num_of_contr_assigns`` by 1.
9. Lets add to our if/else statement. Create two variables ``num_of_trt_status_success`` and ``num_of_ctr_status_success`` and assign them both the value 0. Inside our loop, when a patient treatment is "t" we will implement step 8 and in addition we will increment the variable ``num_of_trt_status_success`` by 1 if the patient status is a success. If the patient treatment is "c" then we will increment ``num_of_ctr_status_success`` by 1 if the patient status is a success. 
10. Using ``num_of_trt_status_success`` and ``num_of_treat_assigns`` compute the conditional probability of success given the patient was assigned treatment. 
11. Using ``num_of_ctr_status_success`` and ``num_of_contr_assigns`` compute the conditional probability of success given the patient was assigned control. 

In [14]:
patient_status    = c(0,1,1,0,0,1,1,1,0,0,0,1,1,1,0,0,0,0,1,0,0,0,1,1,1,0)
patient_treatment = c("t","c","t","c","t","c","t","t","c","t","c","t","c","t","c","t","c","t","t","c","t", "c","c","t","c","c")

**Understand your data**  
We have two vectors to work with, `patient_status` and `patient_treatment`. We can infer that these vectors represent different aspects of information about the same patients, namely their status (ie, are they healthy `1` or not `0`), and their treatment (did they get treatment `t` or not `c`, control). If `patient_status` and `patient_treatment` correspond to the same set of patients, their length should be equal:

In [15]:
length(patient_status) == length(patient_treatment)

[1] TRUE

It's always a good idea to test these types of assumptions about your data. So this means that each element `i` in both `patient_status` and `patient_treatment` refer to the same patient, $Patient_i$.

**Question 1**  
1. Use the ``length`` function to compute the number of patients in the study and store that length in the variable ``N``  

In [16]:
N = length(patient_status)

**Question 2**  

2. Create a variable called ``num_of_treat_assigns`` and assign that variable the value 0  

To create a variable, we declare the name of the variable, followed by an equals sign, followed by what we want to assign as the value of that variable:

In [17]:
num_of_treat_assigns = 0

**Question 3**  
3. Create a variable called ``num_of_contr_assigns`` and assign that variable the value 0

In [18]:
num_of_contr_assigns = 0

**Question 4**  
4. Create a variable called ``num_of_status_success`` and assign that variable the value 0  

In [19]:
num_of_status_success = 0

**Question 5**  
5. Create a for loop that will run a code block starting at 1 and ending at ``N``

In [20]:
for (i in 1:N) {
    ## code to execute on each iteration
}

**Question 6**  
6. Inside the code block of the for loop, use the variable ``num_of_status_success`` to count the number of patients with a succussful status  

So the `for` loop we've built runs once per each patient. And, for each patient, we want to ask whether that patient's status is a success or not. We can use the counter variable we've created in the `for` loop (`i`) to index the relevant vector (`patient_status`) in order to count up the ones in this vector. Maybe you noticed that we could get this information without a `for` loop, by doing something like: `sum(patient_status)`. This demonstrates how there are many ways to solve programming problems! For the purposes of this exercise, however, we want you to get practice using `for` loops:

In [21]:
for (i in 1:N){
    if (patient_status[i] == 1) {
        num_of_status_success = num_of_status_success + 1
    }
}

**Question 7**  
7. Use ``N`` and ``num_of_status_success`` to compute the probability of a success  

We compute the frequentist probability to solve this question:  

$$
P(Success) = \frac{\text{Number of Things I Care About}}{\text{Total Number of Things}}\\
P(Success) = \frac{\text{Number of Status Successes}}{\text{Total Number of Patients}}\\
P(Success) = \frac{NumOfStatusSuccess}{N}
$$

In [22]:
print(num_of_status_success / N) 

[1] 0.4615385


**Question 8**  
8. Lets make our for loop more complicated. Inside our code block, create an if/else such that when patient treatment is a "t" we increment the variable ``num_of_treat_assigns`` by 1, else we increment the variable ``num_of_contr_assigns`` by 1.  

Now we care about the vector `patient_treatment`, and we need to use an `else` statement along with the `if`:

In [23]:
for (i in 1:N) {
    if (patient_treatment[i] == 't') {
        num_of_treat_assigns = num_of_treat_assigns + 1
    } else {
        num_of_contr_assigns = num_of_contr_assigns + 1
    }
}
print(num_of_treat_assigns)
print(num_of_contr_assigns)

[1] 13
[1] 13


If this data came from an experiment, the number of subjects in a treatment condition should be close if not equal to the number of subjects in a control condition, so these results make sense.

**Question 9**  
9. Lets add to our if/else statement. Create two variables ``num_of_trt_status_success`` and ``num_of_ctr_status_success`` and assign them both the value 0. Inside our loop, when a patient treatment is "t" we will implement step 8 and in addition we will increment the variable ``num_of_trt_status_success`` by 1 if the patient status is a success. If the patient treatment is "c" then we will increment ``num_of_ctr_status_success`` by 1 if the patient status is a success.   

To solve this problem, we want to make use of nested if/else statements. We want to divide patients into four different categories based on their status and treatment: treatment and success, treatment and no success, control and success, control and no success. 

Imagine that you worked in a hospital and you had these patients in a room and you wanted to make these groups. What would you do? You might first split up everyone according to whether or not they got the treatment, and then split them up again according to whether or not their status is a success. That's exactly what we're doing with code by using a first `if` statement to ask about treatment, and then a nested `if` statement to ask about status.

In [26]:
## important to refresh the following two variables to zero
num_of_treat_assigns = 0
num_of_contr_assigns = 0

num_of_trt_status_success = 0 
num_of_ctr_status_success = 0

for (i in 1:N) {
    if (patient_treatment[i] == 't') {
        num_of_treat_assigns = num_of_treat_assigns + 1
        if (patient_status[i] == 1) {
            num_of_trt_status_success = num_of_trt_status_success + 1
        }
    } else {
        num_of_contr_assigns = num_of_contr_assigns + 1 
        if (patient_status[i] == 1) {
            num_of_ctr_status_success = num_of_ctr_status_success + 1
        }
        
    }
}
## these two should equal what we got in the previous question
print(num_of_treat_assigns)
print(num_of_contr_assigns)

print(num_of_trt_status_success)
print(num_of_ctr_status_success)

[1] 13
[1] 13
[1] 7
[1] 5


I also know that the sum of the last two variables should equal the sum of all the ones in the `patient_status` vector. This is because our code above is 'catching' all of those ones and just grouping them according to whether they belong to the treatment or control group.

In [27]:
7 + 5 == sum(patient_status)

[1] TRUE

**Question 10**  
10. Using ``num_of_trt_status_success`` and ``num_of_treat_assigns`` compute the conditional probability of success given the patient was assigned treatment.  

The formula for conditional probability:  

$$
P(A|B) = \frac{P(A \cap B)}{P(B)} \\
P(Success | Treatment) = \frac{P(Success \cap Treatment)}{P(Treatment)}
$$

We can use the frequentist approach to solve for both $P(Treatment)$ and $P(Success \cap Treatment)$:



$$
P(Treatment) = \frac{N_{Treatments}}{N_{Patients}} \\
P(Success \cap Treatment) = \frac{N_{Patients,Treatment,Success}}{N_{Patients}}
$$

In [28]:
p_treatment = num_of_treat_assigns / N
p_success_treatment = num_of_trt_status_success / N

p_success_treatment / p_treatment

[1] 0.5384615

**Question 11**  
11. Using ``num_of_ctr_status_success`` and ``num_of_contr_assigns`` compute the conditional probability of success given the patient was assigned control.  

We use the same logic as above, first using the frequentist approach to estimate $P(Control)$ and $P(Control \cap Success)$, then divide these quantities to compute $P(Success | Control)$:

In [29]:
p_control = num_of_contr_assigns / N
p_success_control = num_of_ctr_status_success / N

p_success_control / p_control

[1] 0.3846154

We see that the probability of success given the control ($0.38$) is much lower than the probability of success given the treatment ($0.54$), so we can conclude that the treatment is effective.